File: lib/py/etl/paradicms_etl/models/dc/dc_images_mixin.py
from typing import Tuple

from rdflib import DCTERMS, URIRef

from paradicms_etl.models.image import Image
from paradicms_etl.models.images_mixin import ImagesMixin
from paradicms_etl.models.resource_backed_model_mixin import ResourceBackedModelMixin


class DcImagesMixin(ResourceBackedModelMixin, ImagesMixin):
    class Builder(ResourceBackedModelMixin.Builder, ImagesMixin.Builder):
        def add_image(self, image: URIRef) -> "DcImagesMixin.Builder":
            # (image, dcterms:source, this) in this graph.
            # Dublin Core doesn't have an inverse of dcterms:source.
            if isinstance(image, Image):
                image_uri = image.uri
            else:
                image_uri = image
            self._resource.graph.add(
                (image_uri, DCTERMS.source, self._resource.identifier)
            )
            return self

    @property
    def image_uris(self) -> Tuple[URIRef, ...]:
        return tuple(
            subject
            for subject in self.resource.graph.subjects(
                object=self.resource.identifier, predicate=DCTERMS.source
            )
            if isinstance(subject, URIRef)
        )

File: lib/ts/models/src/dc/DcImagesMixin.ts
import {dcterms} from "@paradicms/vocabularies";
import {Memoize} from "typescript-memoize";
import {Image} from "../Image";
import {ImagesMixin} from "../ImagesMixin";
import {ModelIdentifier} from "../ModelIdentifier";
import {ResourceBackedModelMixin} from "../ResourceBackedModelMixin";
import {mapTermToImage} from "../mapTermToImage";

export abstract class DcImagesMixin extends ResourceBackedModelMixin
  implements ImagesMixin {
  @Memoize()
  get images(): readonly Image[] {
    // (image, dcterms:source, this).
    // Dublin Core doesn't have an inverse of dcterms:source like FOAF has foaf:depiction and foaf:depicts.
    const images: Image[] = [];
    for (const quad of this.dataset.match(
      null,
      dcterms.source,
      this._identifier
    )) {
      const image = mapTermToImage(this, quad.subject as ModelIdentifier);
      if (image) {
        images.push(image);
      }
    }
    return images;
  }

  protected preMemoizeImages() {
    this.images;
  }
}
